### PR TITLE
0009891: Add support for more map attributes

### LINKED ISSUE #454: Add missing map attributes
**Describe the solution you'd like**
Many attributes have not been added and could be useful for mapping with no longer a need of having a script to manipulate them afterwards. Fairly small to add.

**Additional context**
https://bugs.mtasa.com/view.php?id=9891
https://github.com/multitheftauto/mtasa-resources/pull/95
https://github.com/multitheftauto/mtasa-blue/pull/263

### PR DESCRIPTION
**Pull request by patrikjuvonen**
**Mantis Bug Tracker issue:**
[9891](https://bugs.multitheftauto.com/view.php?id=9891)
fixes #454 

**Summary:**
- Add _visibleDistance_ and _size_ attributes to `<blip>`
- Add _headless_ and _walkingStyle_ attributes to `<ped>`
- Make "all" option case-insensitive for _upgrades_ attribute on `<vehicle>`
- Add _headLightColor_, _taxiLightOn_, _engineOn_, _lightsOn_, _damageProof_, _explodableFuelTank_, _toggleRespawn_, _respawnDelay_, _respawnPosX_, _respawnPosY_, _respawnPosZ_, _respawnRotX_, _respawnRotY_, _respawnRotZ_ attributes to `<vehicle>`
- Make code style more consistent
- mtasa-resources edf + editor update issue + PR open here: https://github.com/multitheftauto/mtasa-resources/issues/115

**Test map file:**
[test_263_v2.txt](https://github.com/multitheftauto/mtasa-blue/files/4239664/test_263_v2.txt)

### CODE REVIEW

File: Server/mods/deathmatch/logic/CBlip.cpp [hunk 1]
```diff
@@ -91,31 +92,18 @@ bool CBlip::ReadSpecialData(void)
...
     // Grab the "ordering" data
     if (GetCustomDataInt("ordering", iTemp, true))
-    {
-        if (iTemp >= -32768 && iTemp <= 32767)
-        {
-            m_sOrdering = static_cast<short>(iTemp);
-        }
-        else
-        {
-            CLogger::ErrorPrintf("Bad 'ordering' id specified in <blip> (line %u)\n", m_uiLine);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CBlip.cpp]: bad ordering error is now suppressed.

File: Server/mods/deathmatch/logic/CBlip.cpp [hunk 2]
```diff
@@ -91,31 +92,18 @@ bool CBlip::ReadSpecialData(void)
             return false;
         }
     }
-    else
-    {
-        m_Color = SColorRGBA(0, 0, 255, 255);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CBlip.cpp:113]: why was this moved up?

File: Server/mods/deathmatch/logic/CPed.cpp [hunk 3]
```diff
@@ -190,28 +192,36 @@ bool CPed::ReadSpecialData(void)
...
-    if (GetCustomDataBool("frozen", bFrozen, true))
-        m_bFrozen = bFrozen;
+    // Grab the "frozen" data
+    GetCustomDataBool("frozen", m_bFrozen, true);
+
+    // Grab the "headless" data
+    GetCustomDataBool("headless", m_bHeadless, true);
+
+    // Grab the "walkingStyle" data
+    GetCustomDataInt("walkingStyle", m_iMoveAnim, true);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CPed.cpp:242]: to confirm, if this is not in the map, it will set it to 0, right?

File: Server/mods/deathmatch/logic/CVehicle.cpp [hunk 4]
```diff
@@ -245,67 +245,46 @@ bool CVehicle::ReadSpecialData(void)
...
-    {
         m_usAdjustableProperty = 0;
-    }
 
     // Grab the "color" data
     char szTemp[256];
     if (GetCustomDataString("color", szTemp, 256, true))
     {
-        uchar ucValues[12] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+        uchar ucValues[12] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CVehicle.cpp]: unwanted style change

File: Server/mods/deathmatch/logic/CVehicle.cpp [hunk 5]
```diff
@@ -378,8 +404,8 @@ bool CVehicle::ReadSpecialData(void)
     // Grab the "damageProof" data
     GetCustomDataBool("damageProof", m_bDamageProof, true);
 
-    // Grab the "fuelTankExplodable" data
-    GetCustomDataBool("fuelTankExplodable", m_bFuelTankExplodable, true);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CVehicle.cpp]: potential breaking change?

File: Server/mods/deathmatch/logic/Utils.cpp [hunk 6]
```diff
@@ -526,18 +526,97 @@ bool XMLColorToInt(const char* szColor, unsigned long& ulColor)
...
+        ucRed = vecColors[0].R;
+        ucGreen = vecColors[0].G;
+        ucBlue = vecColors[0].B;
+        ucAlpha = vecColors[0].A;
+        return true;
+    }
+    return false;
+}
+
+bool ColorStringToRGBA(const char* szColor, SColorRGBA defaultColor, std::vector<SColorRGBA>& vecColors, unsigned char& ucCount, bool bIgnoreAlpha)
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/Utils.cpp]: Please can you explain what "mixed colours" are? This should land in a separate pull request as it's not directly related to supporting more map attributes.

**Review by qaisjp — DISMISSED** [review]
Thanks for this. Lots of changes in this one, so it's fairly difficult to review!

I don't think this one will land in 1.5.6, but I've reviewed it now nonetheless.

File: Server/mods/deathmatch/logic/CBlip.cpp (hunk 1 above)
Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CBlip.cpp]: That's because the final value gets clamped, so there is no bad order anymore.

File: Server/mods/deathmatch/logic/CBlip.cpp (hunk 2 above)
Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CBlip.cpp:113]: Because it would always fall back to that anyway. Does it cause issues?

File: Server/mods/deathmatch/logic/CPed.cpp (hunk 3 above)
Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CPed.cpp:242]: Correct.

File: Server/mods/deathmatch/logic/Utils.cpp (hunk 6 above)
Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/Utils.cpp]: Sure, I can separate it.

By mixed colors I mean that we can have hexadecimal and RGB values mixed up and it will be able to handle it. So let's say you didn't have the RGB color values for a specific red, but you have the hexadecimal color value, you can input that even if you have other RGB color values in the string.

Example: `255, 255, 255, #ff0000, 0, 0, 0, #fff`
This will cause color 1 to be white, color 2 to be red, color 3 to be black, color 4 to be white.
It essentially just speeds up the process so you don't have to look up the RGB values for an already valid hexadecimal color value you have, but is not supported at the moment. Perhaps a quality of life tweak.

At the moment the problem is that you can only use RGB colors for vehicle colors, and you can only use hexadecimal colors for anything else (blips, teams, radar areas, headlight colors). So this changes it to so that you can use RGB on anything else without having to rewrite everything all over again. Essentially adding support for hexadecimal and RGB color values at the same time for the same utility method.

File: Server/mods/deathmatch/logic/CVehicle.cpp (hunk 5 above)
Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CVehicle.cpp]: That's a commit I made that I reverted. Doesn't exist in MTA at the moment. Named it `explodableFuelTank` because I think that sounds better, but whichever is fine.

File: Server/mods/deathmatch/logic/CVehicle.cpp (hunk 4 above)
Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CVehicle.cpp]: Done.

File: Server/mods/deathmatch/logic/CBlip.cpp (hunk 1 above)
Comment (qaisjp) [reply Server/mods/deathmatch/logic/CBlip.cpp]: Do we clamp values anywhere else? I think in this case it might be good to show an error so that people don't get confused between `999999` and `99999999999` behaving as if it's the same level.

**Comment by qaisjp** [comment]
changes addressed*

**Comment by qaisjp** [comment]
For SetPaletteColors and SetRGBColors, do you think mappers might be using black (or whatever the default is) to mean "random color"?

**Comment by patrikjuvonen** [comment]
I think that should totally not be the case. If you want random color, you don't define the color attribute, and if you want black, you define it black. Right now defining color as black doesn't give you black but a random color.

File: Server/mods/deathmatch/logic/CBlip.cpp (hunk 1 above)
Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CBlip.cpp]: Ok, no problem, I'll add the error back in. I'll also check if we do more clamping elsewhere.

Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CBlip.cpp]: Realized that was some old commit that you referred to. Anyway the error is there right now.

So in this PR:
  - Blip `size` gets clamped between 0 and 25. [Function equivalent](https://wiki.multitheftauto.com/wiki/SetBlipSize) clamps too.
  - Blip `visibleDistance` gets clamped between 0 and 65535. [Function equivalent](https://wiki.multitheftauto.com/wiki/SetBlipVisibleDistance) clamps too.

Should I change these two map attributes to result in an error instead rather than clamping silently?

Comment (qaisjp) [reply Server/mods/deathmatch/logic/CBlip.cpp]: Yes, I think reporting it as an error would be good. I think later we can go over those functions and turn them into warnings.

Comment (patrikjuvonen) [reply Server/mods/deathmatch/logic/CBlip.cpp]: Ok, cheers.

**Comment by patrikjuvonen** [comment]
Only thing left in this PR is to add the random color attribute.

**Comment by patrikjuvonen** [comment]
What we decided on Discord earlier:
- Keep existing color attribute default behavior to keep backwards compatibility.
- Introduce new `forceColor` attribute (default: `false`), which can be used to toggle whether `color` attribute being 0,0,0 will be read as forcefully black or random (default color will be overridden by force). Added for backwards compatibility.
- In the future we may want to make a XSD or other things to define the map files.

I'll do the new attribute today.

**Comment by qaisjp** [comment]
> * Introduce new `forceColor` attribute (default: `false`), which can be used to toggle whether `color` attribute being 0,0,0 will be read as forcefully black or random (default color will be overridden by force). Added for backwards compatibility.

I was thinking:
- the default value of `randomColor` would be `black`, which is the old functionality
- `randomColor=true` would force a random color
- `randomColor=false` would allow you to use the black color

But then it isn't obvious what `randomColor` does by default (I would expect it to be `false` if I was reading a `.map` file).

And my worry with `forceColor` is that it isn't very semantic & has the same issue as above (could someone know what `forceColor=false` does without reading the docs?)

Perhaps we could do a census (ask racing servers to run a script for us and tell us the output), and if results say that `color=0,0,0` is rarely used, we can just choose to break backwards compatibility?

It's up to you really, that sort of census sounds like too much work it would drag this PR on for too long

**Comment by patrikjuvonen** [comment]
I think the problem here is that by introducing `randomColor` attribute I'm expecting to get a random color doesn't matter what I input into the `color` attribute, at least in my opinion. This behavior would break existing maps.

Because it's actually quite "ambiguous" what the default behavior is, by introducing a `forceColor` attribute, we will make it clear that whatever our vehicle class does now or in the future by default will be overridden by the `color` attribute.

So whichever way we go, people would have to read up on Wiki what it does.

We could however name it `__backwardsCompatibleColorBehavior`, `__useOldColorBehavior` (default `true`) or something other specifically dirty so it marks that this isn't good behavior and may be changed in the future to only work with the new style which is that 0,0,0 will result in black color, not random color.

I do however believe we should introduce a new `randomColor` attribute that will generate a random color even if the current default vehicle class behavior was refactored and random color wouldn't happen by default. We could even make it so that if no `color` attribute exists on the node, it will generate a random color, however I think this behavior would break more maps.

And meanwhile, we could inquire race server owners whether they have any `color="0, 0, 0"` strings in any of their map files, then replacing all of these occurrences by `randomColor="true"`, if their intended behavior was indeed to get a random color, and not black color.

**Comment by Woovie** [comment]
I contacted AfuSensi on Discord, who owns/operates/works on Mr. Green's

> `Search "<vehicle.*color="0,0,0" (1485 hits in 167 files)`

> `Search "<vehicle.*color="([1-255]{1,3}),\1,\1" (82 hits in 33 files)`

Around 2500 maps total, and this is the total amount of returns.

**Comment by patrikjuvonen** [comment]
Thanks @Woovie, I think we should separate the vehicle color fix as its own PR and issue so we can discuss that matter in a different thread to see what would be the best solution. I do believe something has to be done since the current default behavior is not expected behavior in my opinion.

**Comment by Woovie** [comment]
So we should revert 274d9b38b74b1482c830c1677b56a7298cd82424 and I guess it's ready for merge? @qaisjp what do you think?

**Comment by qaisjp** [comment]
> So we should revert [274d9b3](https://github.com/multitheftauto/mtasa-blue/commit/274d9b38b74b1482c830c1677b56a7298cd82424) and I guess it's ready for merge? @qaisjp what do you think?

sounds fine

**Comment by patrikjuvonen** [comment]
I have now removed the forced color attribute from the PR. Making a new issue about it now, and also of the broader color syntax seen earlier in this PR.

Black RGB color attribute issue created at #1257 
Hex and RGBA color attribute support issue created at #1258

**Review by qaisjp — APPROVED** [review]

**Comment by qaisjp** [comment]
go for merge, lgtm, :shipit:, @patrikjuvonen
